juju/terraform-provider-juju
# [JUJU-1899] Ensure compatibility w/ `2.9.33` through `3.0-beta`

One issue we faced running the provider with `juju 3.0` was that the `GetModelConstraints()` is no longer on the `Client` api. It's copied into the `ModelConfig` api as part of the cleanups going from juju `2.9` to `3.0`.

This change updates the juju in the `go.mod` to the point of that copy, so the provider can call the `ModelConfig.GetModelConstraints()` regardless of whichever juju is being used. Unfortunately because of this, we no longer support `juju <=2.9.32`. This change also updates the docs with that requirement.

#### QA Steps

To fully qa this, a `2.9.33` and a `3.0-beta4` controllers (bootstrapped on microk8s - or any other k8s substrate) are needed. I'll just paste the `demo.tf` I used to test this. And the steps for an individual run should be something like this:

```
 $ go build
 $ cp terraform-provider-juju <path-to-terraform>/.terraform.d/plugins/registry.terraform.io/juju/juju/0.4.1/linux_amd64/terraform-provider-juju_v0.4.1
 $ cd <path-to-wherever-you-put-the demo.tf>
 $ terraform init
 $ terraform plan
 $ TF_LOG=DEBUG terraform apply
```

So this should succeed in creating the models/applications/relations on both controllers without any errors.

```tf
terraform {
  required_providers {
    juju = {
      source  = "juju/juju"
      version = "0.4.1"
    }
  }
}
provider "juju" {}

# Create a model and deploy Zinc
resource "juju_model" "zinc" {
  name = "zinc"
}

resource "juju_application" "zinc" {
  name  = "zinc"
  model = juju_model.zinc.name
  trust = true

  charm {
    name = "zinc-k8s"
    series = "focal"
  }
}

# Create the monitoring model
resource "juju_model" "monitoring" {
  name = "monitoring"
}

# Deploy Grafana, Prometheus, Loki, Parca
resource "juju_application" "grafana" {
  name  = "grafana"
  model = juju_model.monitoring.name
  trust = true

  charm {
    name    = "grafana-k8s"
    channel = "beta"
    series = "focal"
  }
}

resource "juju_application" "prometheus" {
  name  = "prometheus"
  model = juju_model.monitoring.name
  trust = true

  charm {
    name    = "prometheus-k8s"
    channel = "beta"
    series = "focal"
  }
}

resource "juju_application" "loki" {
  name  = "loki"
  model = juju_model.monitoring.name
  trust = true

  charm {
    name    = "loki-k8s"
    channel = "beta"
    series = "focal"
  }
}

resource "juju_application" "parca" {
  name  = "parca"
  model = juju_model.monitoring.name
  trust = true

  charm {
    name    = "parca-k8s"
    channel = "edge"
    series = "focal"
  }
}

# Monitoring model internal relations
resource "juju_integration" "prometheus-grafana" {
  model = juju_model.monitoring.name

  application {
    name     = juju_application.prometheus.name
    endpoint = "grafana-source"
  }

  application {
    name     = juju_application.grafana.name
    endpoint = "grafana-source"
  }
}

resource "juju_integration" "parca-prometheus" {
  model = juju_model.monitoring.name

  application {
    name = juju_application.prometheus.name
  }

  application {
    name = juju_application.parca.name
  }
}

resource "juju_integration" "parca-grafana" {
  model = juju_model.monitoring.name

  application {
    name = juju_application.grafana.name
  }

  application {
    name = juju_application.parca.name
  }
}

resource "juju_integration" "loki-grafana" {
  model = juju_model.monitoring.name

  application {
    name = juju_application.loki.name
  }

  application {
    name     = juju_application.grafana.name
    endpoint = "grafana-source"
  }
}

# Create offers
resource "juju_offer" "prometheus-offer" {
  model            = juju_model.monitoring.name
  application_name = juju_application.prometheus.name
  endpoint         = "metrics-endpoint"
}

resource "juju_offer" "grafana-offer" {
  model            = juju_model.monitoring.name
  application_name = juju_application.grafana.name
  endpoint         = "grafana-dashboard"
}

resource "juju_offer" "parca-offer" {
  model            = juju_model.monitoring.name
  application_name = juju_application.parca.name
  endpoint         = "profiling-endpoint"
}

resource "juju_offer" "loki-offer" {
  model            = juju_model.monitoring.name
  application_name = juju_application.loki.name
  endpoint         = "logging"
}

# Create the cross-model integrations
resource "juju_integration" "prometheus-zinc" {
  model = juju_model.zinc.name

  application {
    name = juju_application.zinc.name
  }

  application {
    offer_url = juju_offer.prometheus-offer.url
  }
}

resource "juju_integration" "grafana-zinc" {
  model = juju_model.zinc.name

  application {
    name = juju_application.zinc.name
  }

  application {
    offer_url = juju_offer.grafana-offer.url
  }
}

resource "juju_integration" "loki-zinc" {
  model = juju_model.zinc.name

  application {
    name = juju_application.zinc.name
  }

  application {
    offer_url = juju_offer.loki-offer.url
  }
}

resource "juju_integration" "parca-zinc" {
  model = juju_model.zinc.name

  application {
    name = juju_application.zinc.name
  }

  application {
    offer_url = juju_offer.parca-offer.url
  }
}
```
